apps/sms/views/shared/js/utils.js

```diff
@@ -33,40 +33,91 @@
     },
     getFormattedHour: function ut_getFormattedHour(time) {
       this.date.shared.setTime(+time);
-      return this.date.format.localeFormat(
-        this.date.shared,
-        navigator.mozL10n.get(
-          navigator.mozHour12 ? 'shortTimeFormat12' : 'shortTimeFormat24'
-        )
-      );
+      return this.date.shared.toLocaleString(navigator.languages, {
+        hour12: navigator.mozHour12,
+        hour: 'numeric',
+        minute: 'numeric',
+      });
     },
     getDayDate: function re_getDayDate(time) {
       this.date.shared.setTime(+time);
       this.date.shared.setHours(0, 0, 0, 0);
       return this.date.shared.getTime();
     },
-    getHeaderDate: function ut_giveHeaderDate(time) {
-      var _ = navigator.mozL10n.get;
+    _getFormatter: function ut_getFormatter({options, withTime}) {
+      if (withTime) {
+        options.hour12 = navigator.mozHour12;
+        options.hour = 'numeric';
+        options.minute = 'numeric';
+      }
+      return new Intl.DateTimeFormat(navigator.languages, options);
+    },
+    setHeaderDate: function ut_setHeaderDate({time, element, withTime}) {
+      var formatter;
       var today = Utils.getDayDate(Date.now());
       var otherDay = Utils.getDayDate(time);
       var dayDiff = (today - otherDay) / 86400000;
       this.date.shared.setTime(+time);
 
       if (isNaN(dayDiff)) {
-        return _('incorrectDate');
+        element.setAttribute('data-l10n-id', 'incorrectDate');
+        return;
       }
 
       if (dayDiff < 0) {
         // future time
-        return this.date.format.localeFormat(
-          this.date.shared, '%x'
-        );
+        formatter = this._getFormatter({
+          options: {
+            month: '2-digit',
+            day: '2-digit',
+            year: 'numeric',
+          },
+          withTime: withTime
+        });
+        element.removeAttribute('data-l10n-id');
+        element.textContent = formatter.format(this.date.shared);
+        return;
       }
 
-      return dayDiff === 0 && _('today') ||
-        dayDiff === 1 && _('yesterday') ||
-        dayDiff < 6 && this.date.format.localeFormat(this.date.shared, '%A') ||
-        this.date.format.localeFormat(this.date.shared, '%x');
```

Comment:
The bottom line is, I don't think there's any room for a bug in Intl implementation around this. That's coming straight from ICU. There may be bugs in ICU (and we're updating our ICU fairly often (bug 992079 and bug 1075758), but there shouldn't be too many, and with Safari, Chrome, Windows and MacOS all using the same base technology as us, we should get more consistent and error-free experience.

One benefit of this change is that we're moving from localizer-defined formatting to developer-defined formatting. Which means that if next month you'll decide to switch something from 2-digit to numeric or long, you're free to do that at no cost to localizers.

And since we're not close to any release, we'll have a lot of time to ensure the formatting looks good.